state/block.go

```diff
@@ -102,19 +115,28 @@ func (bs *blockState) GetLatestBlock() types.BlockHeader {
 func (bs *blockState) GetBlockByHash(hash common.Hash) (types.Block, error) {
 	header, err := bs.GetHeader(hash)
 	if err != nil {
-		return types.Block{}, nil
+		return types.Block{}, err
 	}
 	blockData, err := bs.GetBlockData(hash)
 	if err != nil {
-		return types.Block{}, nil
+		return types.Block{}, err
 	}
 	blockBody := blockData.Body
 	return types.Block{Header: header, Body: blockBody}, nil
 }
 
-func (bs *blockState) GetBlockByNumber(n *big.Int) types.Block {
-	// Can't do yet
-	return types.Block{}
+func (bs *blockState) GetBlockByNumber(n *big.Int) (types.Block, error) {
+	// First retrieve the block hash based on the block number from the database
+	hash, err := bs.db.Db.Get(headerHashKey(n.Uint64()))
+
+	if err != nil {
+		return types.Block{}, err
+	}
+
+	// Then find the block based on the hash
+	endHash := common.NewHash(hash)
```

Comment:
Is `endHash` a typo?

Reply:
Maybe rename this `hash`, then rename the current `hash` to `data` or `bytes`.

Reply:
done